### PR TITLE
Implementing matrix strategy to speed up gha-only tests

### PR DESCRIPTION
**Pull request by seankross**
Sorry for splitting hairs but I feel like this is the better implementation practice compared to #22. Looking for approval from both @tefirman and @sckott before merging. Also sorry for the sloppy merging. If this gets approval I will close #22.

### CODE REVIEW

File: .github/workflows/cromwell-test-run.yml [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
 
 jobs:
   testrun:
+    if: ${{ github.event.pull_request.head.repo.fork == false }}
     runs-on: ubuntu-latest
     permissions: write-all
+    strategy:
+      matrix:
+        wdl:
+          - helloHostname/helloHostname.wdl
```
Comment (sckott) [inline .github/workflows/cromwell-test-run.yml:15]: i don't know the answer here, just asking. do you want to make it explicit for each file tested (so edit this and other yml file whenever a new wdl folder is added to tests) or maybe use a glob pattern? hmm although maybe that's not possible to do a glob thing

File: .github/workflows/cromwell-test-run.yml [hunk 2]
```diff
@@ -22,11 +30,5 @@ jobs:
...
       -
         name: Execute Test Run of WDL Workflows
-        run: |
-          java -jar cromwell-86.jar run helloHostname/helloHostname.wdl
-          java -jar cromwell-86.jar run helloModuleHostname/helloModuleHostname.wdl
-          java -jar cromwell-86.jar run helloDockerHostname/helloDockerHostname.wdl
-
-
-
+        run: java -jar cromwell-86.jar run ${{ matrix.wdl }}
```
Comment (sckott) [inline .github/workflows/cromwell-test-run.yml:33]: i forget, when does `run`  exit? submits and then exits, or waits for a certain status before exioting?

**Review by sckott — COMMENTED** [review]
a few comments

File: .github/workflows/cromwell-test-run.yml (hunk 2 above)
Comment (tefirman) [reply .github/workflows/cromwell-test-run.yml]: `run` exits once the workflow hits a final state, whether that's success or failure (can't remember the exact status wording of those two though...)

Comment (sckott) [reply .github/workflows/cromwell-test-run.yml]: ok thanks. just curious

Comment (seankross) [reply .github/workflows/cromwell-test-run.yml]: > submits and then exits, or waits for a certain status before exiting?

It waits for sure, if the command returns not `0` then it fails the whole action. You have to add a special arg to allow it to proceed in the event of a failure

File: .github/workflows/cromwell-test-run.yml (hunk 1 above)
Comment (seankross) [reply .github/workflows/cromwell-test-run.yml]: I'm working on this rn

Comment (tefirman) [reply .github/workflows/cromwell-test-run.yml]: Oooo, I like this idea, mind if I give it a shot?

On a related note, I added some automation in #32 that now requires each unit test to have the wdl script follow the same name as the directory and requires an `inputs.json` and `options.json`.

Comment (tefirman) [reply .github/workflows/cromwell-test-run.yml]: Awesome, thanks @seankross !

Comment (seankross) [reply .github/workflows/cromwell-test-run.yml]: > Oooo, I like this idea, mind if I give it a shot?

Hey, sorry I didn't see this until now.

File: .github/workflows/cromwell-test-run.yml [hunk 3]
```diff
@@ -1,32 +1,56 @@
...
+      - name: Find WDL Directories
+        id: set-matrix
+        run: |
+          # Find all directories containing .wdl files
+          dirs=$(find . -type f -name "*.wdl" -exec dirname {} \; | sort -u | sed 's|^\./||' | jq -R . | jq -s | jq -c .)
+          echo "Directories: $dirs"
+          echo "matrix=$dirs" >> $GITHUB_OUTPUT
+
   testrun:
+    needs: generate-matrix
```
Comment (tefirman) [inline .github/workflows/cromwell-test-run.yml:25]: Looks like this is missing from the womtools yml, might be what's causing the issue..

File: .github/workflows/womtools-validate.yml [hunk 4]
```diff
@@ -1,15 +1,36 @@
...
+      - name: Find WDL Directories
+        id: set-matrix
+        run: |
+          # Find all directories containing .wdl files
+          dirs=$(find . -type f -name "*.wdl" -exec dirname {} \; | sort -u | sed 's|^\./||' | jq -R . | jq -s | jq -c .)
+          echo "Directories: $dirs"
+          echo "matrix=$dirs" >> $GITHUB_OUTPUT
+
   womtoolval:
+    if: ${{ github.event.pull_request.head.repo.fork == false }}
```
Comment (tefirman) [inline .github/workflows/womtools-validate.yml:29]: I think you need to add the `needs: generate-matrix` here.

Comment (seankross) [reply .github/workflows/womtools-validate.yml:26]: absolutely, thank you

**Review by sckott — APPROVED** [review]
sweet, run times are looking good, and seems like total time won't get much bigger when adding more wdls given the matrix approach

**Review by tefirman — APPROVED** [review]
Great call on making the `inputs.json` and `options.json` optional. Looks great, thanks @seankross !

**Pull request merged**